### PR TITLE
Fully atomic rocksDB writes

### PR DESCRIPTION
**Pull request by ljedrz**
While we're already using rocksDB's atomic write batches, they are currently only atomic within a single `DataMap`. In order to have fully atomic writes, we need to adjust the current logic to keep track of the queued operations across all the maps and perform a write only once per atomic write batch.

Unlike with the typed `DataMap`s, this forces us to keep track of the untyped/serialized operations within the shared storage object, and to update it in tandem with operations on all the maps. This is made more complicated by the checkpointing logic, but it can be achieved by keeping track of the atomic scope depth (i.e. the checkpoint index) and using a "stack of stacks" instead of a single stack (like in the `DataMap`). This way we are able to keep track of the checkpoints across all the `DataMap`s, and gain improved tracking of nested storage operations.

In order to better reflect the complexity of our typical storage operations, 2 new tests containing nested `DataMap`s are also added.

As a drive-by, the `checkpoint` field in `DataMap` is renamed to `checkpoints` in order to better reflect that it contains a number of them as opposed to a single one at a time. Also, some of the occurrences of replacing a vector with a new empty one are replaced with a call to `Vec::clear` in order to preserve the earlier allocation(s).

### CODE REVIEW

File: ledger/store/src/helpers/rocksdb/internal/map.rs [hunk 1]
```diff
@@ -173,26 +175,33 @@ impl<
...
+        self.checkpoints.lock().clear();
         // Set the atomic batch flag to `false`.
         self.batch_in_progress.store(false, Ordering::SeqCst);
 
+        // Subtract the atomic depth index.
+        let previous_atomic_depth = self.database.atomic_depth.fetch_sub(1, Ordering::SeqCst);
+
+        // If we're at depth 0, it is the final call to `finish_atomic` and the
+        // atomic write batch can be physically executed.
+        if previous_atomic_depth == 1 {
```
Comment (raychu86) [inline ledger/store/src/helpers/rocksdb/internal/map.rs:208]: Do we need to check the edge case if `self.database.atomic_depth` is already zero? Since `fetch_sub` does a wrap if we subtract from 0

File: ledger/store/src/helpers/rocksdb/internal/map.rs [hunk 2]
```diff
@@ -351,13 +360,178 @@ mod tests {
...
     };
-    use console::{account::Address, network::Testnet3, prelude::*};
+    use anyhow::anyhow;
+    use console::{
+        account::{Address, FromStr},
+        network::Testnet3,
+    };
 
     use serial_test::serial;
     use tracing_test::traced_test;
```
Comment (raychu86) [inline ledger/store/src/helpers/rocksdb/internal/map.rs]: Nit: Move the anyhow import lower.

```suggestion
    use console::{
        account::{Address, FromStr},
        network::Testnet3,
    };

    use anyhow::anyhow;
    use serial_test::serial;
    use tracing_test::traced_test;
```

File: ledger/store/src/helpers/rocksdb/internal/map.rs [hunk 3]
```diff
@@ -109,55 +111,55 @@ impl<
...
+        self.atomic_batch.lock().clear();
         // Clear the checkpoint stack.
-        *self.checkpoint.lock() = Default::default();
+        self.checkpoints.lock().clear();
         // Set the atomic batch flag to `false`.
         self.batch_in_progress.store(false, Ordering::SeqCst);
+        // Clear the database-wide atomic batch.
+        self.database.atomic_batch.lock().clear();
+        // Reset the atomic batch depth.
+        self.database.atomic_depth.store(0, Ordering::SeqCst);
```
Comment (raychu86) [inline ledger/store/src/helpers/rocksdb/internal/map.rs:155]: I believe setting the `atomic_depth` to 0 here is correct, but is there ever a case where we should be decrementing by 1, and then ensuring on the last `abort_atomic` call, that we hit 0? Meaning we've properly aborted all maps?

**Comment by ljedrz** [comment]
Pushed a new commit with a simpler setup, as suggested by @raychu86; instead of being a "parallel" database-wide plumbing, it heavily leverages the underlying `DataMap`s, and only requires the tracking of the current atomic batch scope depth in order to trigger with the very last call to `finish_atomic`. This solution is also faster and more lightweight.

**Review by raychu86 — CHANGES_REQUESTED** [review]

File: ledger/store/src/helpers/rocksdb/internal/map.rs (hunk 1 above)
Comment (ljedrz) [reply ledger/store/src/helpers/rocksdb/internal/map.rs:208]: I can add a relevant assertion; all the calls to `start_atomic` should have corresponding calls to `finish_atomic`, so the final value should always be 0.

File: ledger/store/src/helpers/rocksdb/internal/map.rs [hunk 4]
```diff
@@ -109,55 +111,55 @@ impl<
...
+        self.atomic_batch.lock().truncate(checkpoint);
     }
 
     ///
     /// Aborts the current atomic operation.
     ///
     fn abort_atomic(&self) {
         // Clear the atomic batch.
-        *self.atomic_batch.lock() = Default::default();
+        self.atomic_batch.lock().clear();
```
Comment (raychu86) [inline ledger/store/src/helpers/rocksdb/internal/map.rs:147]: Just curious: Is there a benefit to using `clear` vs reassigning to `Default::default()`?

Comment (ljedrz) [reply ledger/store/src/helpers/rocksdb/internal/map.rs:147]: yes; this reuses the previous allocation, which is very likely of an appropriate size

File: ledger/store/src/helpers/rocksdb/internal/map.rs (hunk 3 above)
Comment (ljedrz) [reply ledger/store/src/helpers/rocksdb/internal/map.rs:155]: `abort_atomic` used to "bubble up" in the past, but this has recently changed; now, instead of being called in case of errors, it is only called in `atomic_finalize`, after all the other operations; I _think_ we should be fine decrementing it as well, but I think it would require a multi-map test for `atomic_finalize`

File: ledger/store/src/helpers/rocksdb/internal/map.rs [hunk 5]
```diff
@@ -90,6 +89,9 @@ impl<
     fn start_atomic(&self) {
         // Set the atomic batch flag to `true`.
         self.batch_in_progress.store(true, Ordering::SeqCst);
+        // Increment the atomic depth index.
+        self.database.atomic_depth.fetch_add(1, Ordering::SeqCst);
+
```
Comment (raychu86) [inline ledger/store/src/helpers/rocksdb/internal/map.rs:94]: We may want to check that the underlying `atomic_batch` in `self.database` is also empty.

```suggestion
// Ensure that the database atomic batch is empty.
assert!(self.database.atomic_batch.lock().is_empty());
```

File: ledger/store/src/helpers/rocksdb/internal/map.rs [hunk 6]
```diff
@@ -109,55 +113,55 @@ impl<
...
-        // Acquire the write lock on the atomic batch.
-        let mut atomic_batch = self.atomic_batch.lock();
-
         // Retrieve the last checkpoint.
-        let checkpoint = self.checkpoint.lock().pop().unwrap_or(0);
-
+        let checkpoint = self.checkpoints.lock().pop().unwrap_or(0);
         // Remove all operations after the checkpoint.
-        atomic_batch.truncate(checkpoint);
+        self.atomic_batch.lock().truncate(checkpoint);
```
Comment (howardwu) [inline ledger/store/src/helpers/rocksdb/internal/map.rs]: TODO: Revert this change.

File: ledger/store/src/helpers/rocksdb/internal/map.rs [hunk 7]
```diff
@@ -173,26 +177,37 @@ impl<
...
+        // calls to `start_atomic` have corresponding calls to `finish_atomic`.
+        assert!(previous_atomic_depth != 0);
+
+        // If we're at depth 0, it is the final call to `finish_atomic` and the
+        // atomic write batch can be physically executed.
+        if previous_atomic_depth == 1 {
+            // Empty the collection of pending operations.
+            let batch = mem::take(&mut *self.database.atomic_batch.lock());
+            // Execute all the operations atomically.
+            self.database.rocksdb.write(batch)?;
```
Comment (howardwu) [inline ledger/store/src/helpers/rocksdb/internal/map.rs:212]:
```suggestion
            self.database.rocksdb.write(batch)?;
            // Ensure that the database atomic batch is empty.
            assert!(self.database.atomic_batch.lock().is_empty());
```

File: ledger/store/src/helpers/rocksdb/internal/map.rs [hunk 8]
```diff
@@ -1071,21 +1257,27 @@ mod tests {
...
+
                     // Update the key.
                     map.insert(0, "2".to_string()).unwrap();
 
                     Ok(())
                 });
                 assert!(result.is_ok());
 
+                // Make sure the checkpoint index is 1.
+                assert_eq!(map.checkpoints.lock().last(), Some(&1));
```
Comment (howardwu) [inline ledger/store/src/helpers/rocksdb/internal/map.rs:1281]:
```suggestion
                assert_eq!(map.checkpoints.lock().last(), Some(&1));
                // Ensure that the atomic batch is empty.
                assert!(self.atomic_batch.lock().is_empty());
                // Ensure that the database atomic batch size is 1.
                assert_eq!(self.database.atomic_batch.lock().len(), 1);
                // Ensure that the database atomic depth is 1.
                assert_eq!(self.database.atomic_depth.load(Ordering::SeqCst), 1);
```

**Review by howardwu — APPROVED** [review]
LGTM

**Review by raychu86 — APPROVED** [review]
LGTM
